drivers/src/tla2528.cpp

```diff
@@ -0,0 +1,137 @@
+    hal::bit_modify(m_pin_cfg).clear(channel_mask);
+  } else {  // must be pin_mode::digitalInput
+    hal::bit_modify(m_pin_cfg).set(channel_mask);
+    hal::bit_modify(m_gpio_cfg).clear(channel_mask);
+  }
+  std::array<hal::byte, 7> cmd_buffer = { op_codes::continuous_register_write,
+                                          register_addresses::pin_cfg,
+                                          m_pin_cfg,
+                                          0x00,
+                                          m_gpio_cfg,
+                                          0x00,
+                                          m_gpo_drive_cfg };
+  hal::write(m_bus, m_i2c_address, cmd_buffer);
+}
+
+void tla2528::set_digital_out(hal::byte p_values)
+{
+  m_gpo_value = p_values;
+  std::array<hal::byte, 3> cmd_buffer = { op_codes::single_register_write,
+                                          register_addresses::gpo_value,
+                                          m_gpo_value };
+  hal::write(m_bus, m_i2c_address, cmd_buffer);
+}
+
+void tla2528::set_digital_out(hal::byte p_channel, bool level)
+{
+  if (p_channel > 7)
+    throw hal::argument_out_of_domain(this);
+  if (level) {
+    hal::bit_modify(m_gpo_value).set(hal::bit_mask::from(p_channel));
+  } else {
+    hal::bit_modify(m_gpo_value).clear(hal::bit_mask::from(p_channel));
+  }
+  set_digital_out(m_gpo_value);
+}
+
+bool tla2528::get_digital_out(hal::byte p_channel)
+{
+  if (p_channel > 7)
+    throw hal::argument_out_of_domain(this);
+  std::array<hal::byte, 1> data_buffer;
+  std::array<hal::byte, 2> cmd_buffer = {
+    op_codes::single_register_read,
+    register_addresses::gpo_value,
+  };
+  hal::write(m_bus, m_i2c_address, cmd_buffer);
+  hal::read(m_bus, m_i2c_address, data_buffer);
+  return hal::bit_extract(hal::bit_mask::from(p_channel), data_buffer[0]);
+}
+
+hal::byte tla2528::get_digital_in()
+{
+  std::array<hal::byte, 1> data_buffer;
+  std::array<hal::byte, 3> cmd_buffer = {
+    op_codes::single_register_read,
+    register_addresses::gpi_value,
+  };
+  hal::write(m_bus, m_i2c_address, cmd_buffer);
+  hal::read(m_bus, m_i2c_address, data_buffer);
+  return data_buffer[0];
+}
+
+bool tla2528::get_digital_in(hal::byte p_channel)
+{
+  if (p_channel > 7)
+    throw hal::argument_out_of_domain(this);
+  return hal::bit_extract(hal::bit_mask::from(p_channel), get_digital_in());
+}
+
+float tla2528::get_analog_in(hal::byte p_channel)
+{
+  set_analog_channel(p_channel);
+  std::array<hal::byte, 2> data_buffer;
+  std::array<hal::byte, 1> read_cmd_buffer = { op_codes::single_register_read };
+
+  hal::write(m_bus, m_i2c_address, read_cmd_buffer);
+  hal::read(m_bus, m_i2c_address, data_buffer);
+
+  // TODO: look into averaging & CRC
+  uint16_t data = 0;
+  hal::bit_modify(data).insert<hal::bit_mask::from(4, 11)>(data_buffer[0]);
+  hal::bit_modify(data).insert<hal::bit_mask::from(0, 3)>(
+    hal::bit_extract(hal::bit_mask::from(4, 7), data_buffer[1]));
+  return data / 4095.0;
```

Comment:
Give the bit masks names so its easier to understand what they are.
Split up the bit_extract into its own named variable so its easier to distinguish against the rest of the code here.

Reply:
updated according to the conversation we had, using manual bit op and added a comment pointing to the datasheet.